src/main.rs

```diff
@@ -33,33 +34,77 @@ struct Args {
     wikipedia_urls: Option<PathBuf>,
 }
 
-fn write(dir: impl AsRef<Path>, page: Page) -> anyhow::Result<()> {
-    let Some(qid) = page.main_entity.map(|e| e.identifier) else {
-        // TODO: handle and still write
-        bail!("Page in list but without wikidata qid: {:?} ({})", page.name, page.url);
+/// Determine the directory to write the article contents to, create it, and create any necessary symlinks to it.
+fn create_article_dir(base: impl AsRef<Path>, page: &Page) -> anyhow::Result<PathBuf> {
+    let Some(qid) = page.wikidata() else {
+        // TODO: Handle and write to wikipedia title directory
+        info!("Page without wikidata qid: {:?} ({})", page.name, page.url);
+        return Ok(Default::default());
     };
 
-    let mut filename = dir.as_ref().to_owned();
-    filename.push(qid);
+    // Otherwise use wikidata as main directory and symlink from wikipedia titles.
+
+    let wikidata_dir = qid.get_dir(base.as_ref().to_owned());
+    create_dir_all(&wikidata_dir)
+        .with_context(|| format!("creating wikidata directory {:?}", &wikidata_dir))?;
+
+    // Write symlinks to main directory.
+    // TODO: Only write redirects that we care about.
+    for title in page.titles() {
+        let title = match title {
+            Err(e) => {
+                warn!("Unable to parse title for page {:?}:  {:#}", page.name, e);
+                continue;
+            }
+            Ok(title) => title,
+        };
+
+        // NOTE: Some wikipedia titles have '/' in them.
+        let wikipedia_dir = title.get_dir(base.as_ref().to_owned());
+        // TODO: handle incorrect links, directories
+        if !wikipedia_dir.exists() && !wikipedia_dir.is_symlink() {
+            let parent_dir = wikipedia_dir.parent().unwrap();
+            create_dir_all(parent_dir)
+                .with_context(|| format!("creating wikipedia directory {:?}", parent_dir))?;
+            symlink(&wikidata_dir, &wikipedia_dir).with_context(|| {
+                format!(
+                    "creating symlink from {:?} to {:?}",
+                    wikipedia_dir, wikidata_dir
+                )
+            })?;
+        }
+    }
+
+    Ok(wikidata_dir)
+}
+
+/// Write selected article to disk.
+///
+/// - Write page contents to wikidata page (`wikidata.org/wiki/QXXX/lang.html`).
+/// - If the page has no wikidata qid, write contents to wikipedia location (`lang.wikipedia.org/wiki/article_title/lang.html`).
```

Comment:
Lang is used two times here in the path, but only one file is always stored in the directory, right?

Reply:
The behavior that the generator/scraper expects is to write all available translations in each directory.
So for the article for [Berlin](https://en.wikipedia.org/wiki/Berlin), if there are OSM tags for `wikipedia:en=Berlin`, `wikipedia:de=Berlin`, `wikipedia:fr=Berlin` and `wikidata=Q64`, and the generator keeps them all, then there will be four folders with duplicates of all language copies:

```
en.wikipedia.org/wiki/Berlin/{en.html, de.html, fr.html, ...}
de.wikipedia.org/wiki/Berlin/{en.html, de.html, fr.html, ...}
fr.wikipedia.org/wiki/Berlin/{en.html, de.html, fr.html, ...}
wikidata/Q64/{en.html, de.html, fr.html, ...}
```

Now, I don't understand exactly how the generator picks which tags to use yet, but just from looking at the Canada Yukon region map there are duplicated copies of wikipedia items there.

For this program, we only see one language at a time, so we write that copy to the master wikidata directory. When later we get the same article in a different language, we write it to the same wikidata directory.

Once all the languages have been processed, it would look like:
```
en.wikipedia.org/wiki/Berlin/ -> wikidata/Q64/
de.wikipedia.org/wiki/Berlin/ -> wikidata/Q64/
fr.wikipedia.org/wiki/Berlin/ -> wikidata/Q64/
wikidata/Q64/{en.html, de.html, fr.html, ...}
```